### ISSUE TITLE
Crash during destruction when using meta_function::new_index and coroutines

### ISSUE THREAD

**Issue by Maddimax** [ThePhD/sol2#1594]
I attached a reproduction example to the ticket: [main.cpp.txt](https://github.com/ThePhD/sol2/files/14865563/main.cpp.txt)

The example crashes during the `m_entries.clear();` in the `~MyDynamic()` destructor.

The crash only occurs if the `test()` function is started as a coroutine.

I sadly don't know enough about lua to see if this is a problem with lua itself, or something in sol2. Any help/hints would be greatly appreciated.

Build and reproduced on macOS 14,

```
clang++ --version                                                                                                                                              e0bb8f0 ✭ ✱ ◼
Apple clang version 15.0.0 (clang-1500.3.9.4)
Target: arm64-apple-darwin23.2.0
Thread model: posix
InstalledDir: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin
```

**Comment by EvanBalster** [comment]
I'm encountering something very similar on Windows 11 with LuaJIT.

This data structure causes no issues except when created and traversed in a coroutine.  The coroutine is managed by Lua code.  When used in a coroutine, it causes semi-random crashing behavior resembling heap corruption.  These crashes happen either in LuaJIT's allocation functions or in the deref function.  In the latter case, it looks like the registry's freelist (a linked list of "free slots" in the registry) is getting corrupted and I believe this is the cause of the other crashes.

Both my code and yours involve a C++ type that contains a list of sol reference types (`reference` in my case, `object` in yours).  These "own" a slot in the registry and free (deref) that slot when destructed — causing the crash.

I'm curious how you got past this problem.  After a day of debugging I understand how I could work around the crash by avoiding the use of registry values (eg, I could replace the C++ type with a Lua table containing all the necessary info and functionality).  But I don't understand why the issue occurs in the first place.

**Comment by EvanBalster** [comment]
Found a lead in [this comment from ThePhD](https://github.com/ThePhD/sol2/issues/890#issuecomment-552064783)...

Coroutines may have their own `lua_State`.  Apparently this means they have their own separate Lua registry, which dies when the coroutine dies.  The `stack_object` from the coroutine references the coroutine's lua state and the `sol::object` created from that stack object lives in the coroutine's state too.  Your `MyDynamic` class gets filled up with objects belonging to a coroutine, which dies, and takes its state and the objects with it.  This invalidates all those `sol::object` you stored, and when your destructor tries to delete them again, it's liable to smash up Lua's allocators.

The workaround for this is for any API that interfaces with coroutines to use `main_object`, `main_reference`, `this_main_state` etc which will avoid using the coroutine's storage.

**Comment by Maddimax** [comment]
You are correct. I had found the issue / solution in the meantime, but must have forgot about this ticket. So closing it as it duplicated #890.
